[Dynamic Region] TiFlash crash for Found existing key in hex

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->
```
[2023/07/21 14:02:45.222 +08:00] [ERROR] [Region.cpp:704] ["[region 328, applied: term 6 index 15697] catch exception: Found existing key in hex: 7480000000000000FF745F7203FC000000FF00028B5F01797955FF7275444F51FF7849FF36594F424D6BFF70FF74536E53457756FFFF7151675A45715048FFFF57554D3469435AFF38FF554267616578FF4A63FF78456E7A6DFF717979FF67677345FF4A56706DFF47316BFF5250513341FF496BFF54316E794D71FF38FF77486F32656839FFFF455A656875746E52FFFF6A5A476B000000FF00FB000000000000F9F9DA2576F76FFE45, while applying `CmdType::Put` on [term 6, index 15698], CF write"] [thread_id=75]
[2023/07/21 14:02:48.854 +08:00] [ERROR] [Exception.cpp:90] ["Code: 49, e.displayText() = DB::Exception: Found existing key in hex: 7480000000000000FF745F7203FC000000FF00028B5F01797955FF7275444F51FF7849FF36594F424D6BFF70FF74536E53457756FFFF7151675A45715048FFFF57554D3469435AFF38FF554267616578FF4A63FF78456E7A6DFF717979FF67677345FF4A56706DFF47316BFF5250513341FF496BFF54316E794D71FF38FF77486F32656839FFFF455A656875746E52FFFF6A5A476B000000FF00FB000000000000F9F9DA2576F76FFE45, e.what() = DB::Exception, Stack trace:\n\n\n       0x1c4994e\tDB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) [tiflash+29661518]\n                \tdbms/src/Common/Exception.h:46\n       0x80a16c1\tDB::RegionCFDataBase<DB::RegionWriteCFDataTrait>::insert(DB::StringObject<true>&&, DB::StringObject<false>&&, DB::DupCheck) [tiflash+134878913]\n                \tdbms/src/Storages/Transaction/RegionCFDataBase.cpp:51\n       0x8099f4b\tauto DB::Region::handleWriteRaftCmd(DB::WriteCmdsView const&, unsigned long, unsigned long, DB::TMTContext&)::$_15::operator()<unsigned long>(unsigned long) const [tiflash+134848331]\n                \tdbms/src/Storages/Transaction/Region.cpp:694\n       0x809875e\tDB::Region::handleWriteRaftCmd(DB::WriteCmdsView const&, unsigned long, unsigned long, DB::TMTContext&) [tiflash+134842206]\n                \tdbms/src/Storages/Transaction/Region.cpp:760\n       0x8053104\tDB::KVStore::handleWriteRaftCmd(DB::WriteCmdsView const&, unsigned long, unsigned long, unsigned long, DB::TMTContext&) const [tiflash+134557956]\n                \tdbms/src/Storages/Transaction/KVStore.cpp:299\n       0x8076405\tHandleWriteRaftCmd [tiflash+134702085]\n                \tdbms/src/Storages/Transaction/ProxyFFI.cpp:97\n  0x7fcb99363dda\t_$LT$engine_store_ffi..observer..TiFlashObserver$LT$T$C$ER$GT$$u20$as$u20$raftstore..coprocessor..QueryObserver$GT$::post_exec_query::h2623763935430ada [libtiflash_proxy.so+25075162]\n  0x7fcb9a3673ed\traftstore::store::fsm::apply::ApplyDelegate$LT$EK$GT$::apply_raft_cmd::hcfec1116af67d6a6 [libtiflash_proxy.so+41866221]\n  0x7fcb9a380df8\traftstore::store::fsm::apply::ApplyDelegate$LT$EK$GT$::process_raft_cmd::hca27e4d70ca755da [libtiflash_proxy.so+41971192]\n  0x7fcb9a386744\traftstore::store::fsm::apply::ApplyDelegate$LT$EK$GT$::handle_raft_committed_entries::hb47bd4b68be5e6d3 [libtiflash_proxy.so+41994052]\n  0x7fcb9a35782c\traftstore::store::fsm::apply::ApplyFsm$LT$EK$GT$::handle_apply::h218c4bca00778724 [libtiflash_proxy.so+41801772]\n  0x7fcb9a35bdb2\traftstore::store::fsm::apply::ApplyFsm$LT$EK$GT$::handle_tasks::hc593b77036cf75f3 [libtiflash_proxy.so+41819570]\n  0x7fcb9944948e\t_$LT$raftstore..store..fsm..apply..ApplyPoller$LT$EK$GT$$u20$as$u20$batch_system..batch..PollHandler$LT$raftstore..store..fsm..apply..ApplyFsm$LT$EK$GT$$C$raftstore..store..fsm..apply..ControlFsm$GT$$GT$::handle_normal::h4c1dcc3342c60fef [libtiflash_proxy.so+26014862]\n  0x7fcb993b4f03\tbatch_system::batch::Poller$LT$N$C$C$C$Handler$GT$::poll::h0867a62a72937d0d [libtiflash_proxy.so+25407235]\n  0x7fcb994af280\tstd::sys_common::backtrace::__rust_begin_short_backtrace::h0abe6a7d2aaa60ee [libtiflash_proxy.so+26432128]\n  0x7fcb994f3d3e\tcore::ops::function::FnOnce::call_once$u7b$$u7b$vtable.shim$u7d$$u7d$::h0b6a956525310e8e [libtiflash_proxy.so+26713406]\n  0x7fcb9ab82d65\tstd::sys::unix::thread::Thread::new::thread_start::hd2791a9cabec1fda [libtiflash_proxy.so+50367845]\n                \t/rustc/96ddd32c4bfb1d78f0cd03eb068b1710a8cebeef/library/std/src/sys/unix/thread.rs:108\n  0x7fcb97849802\tstart_thread [libc.so.6+653314]\n  0x7fcb977e9450\t__clone3 [libc.so.6+259152]"] [source="DB::EngineStoreApplyRes DB::HandleWriteRaftCmd(const DB::EngineStoreServerWrap *, DB::WriteCmdsView, DB::RaftCmdHeader)"] [thread_id=75]
```
### 2. What did you expect to see? (Required)

### 3. What did you see instead (Required)

### 4. What is your TiFlash version? (Required)
TiFlash
Release Version: v7.3.0-alpha
Edition:         Community
Git Commit Hash: a6fcff576c61ee3af11c7f0d6a6f0982866e0914
Git Branch:      heads/refs/tags/v7.3.0-alpha
UTC Build Time:  2023-07-20 11:05:36
Enable Features: jemalloc sm4(GmSSL) avx2 avx512 unwind thinlto
Profile:         RELWITHDEBINFO

Raft Proxy
Git Commit Hash:   1253b471ae6204170fa3917e32e41bac1b4dc583
Git Commit Branch: HEAD
UTC Build Time:    2023-07-20 11:10:58
Rust Version:      rustc 1.67.0-nightly (96ddd32c4 2022-11-14)
Storage Engine:    tiflash
Prometheus Prefix: tiflash_proxy_
Profile:           release
Enable Features:   Unknown (env var does not exist when building)

--- a/dbms/src/Storages/Transaction/Region.cpp
+++ b/dbms/src/Storages/Transaction/Region.cpp
@@ -680,6 +680,8 @@ EngineStoreApplyRes Region::handleWriteRaftCmd(const WriteCmdsView & cmds, UInt6
     Stopwatch watch;
     SCOPE_EXIT({ GET_METRIC(tiflash_raft_apply_write_command_duration_seconds, type_write).Observe(watch.elapsedSeconds()); });
 
+    auto is_v2 = this->getClusterRaftstoreVer() == RaftstoreVer::V2;
+
     const auto handle_by_index_func = [&](auto i) {
         auto type = cmds.cmd_types[i];
         auto cf = cmds.cmd_cf[i];
@@ -691,7 +693,15 @@ EngineStoreApplyRes Region::handleWriteRaftCmd(const WriteCmdsView & cmds, UInt6
             auto tikv_value = TiKVValue(cmds.vals[i].data, cmds.vals[i].len);
             try
             {
-                doInsert(cf, std::move(tikv_key), std::move(tikv_value));
+                if (is_v2)
+                {
+                    // There may be orphan default key in a snapshot.
+                    doInsert(cf, std::move(tikv_key), std::move(tikv_value), DupCheck::AllowSame);
+                }
+                else
+                {
+                    doInsert(cf, std::move(tikv_key), std::move(tikv_value), DupCheck::Deny);
+                }
             }
             catch (Exception & e)
             {

--- a/dbms/src/Storages/Transaction/Region.h
+++ b/dbms/src/Storages/Transaction/Region.h
@@ -216,7 +216,7 @@ private:
 
     // Private methods no need to lock mutex, normally
 
-    void doInsert(ColumnFamilyType type, TiKVKey && key, TiKVValue && value, DupCheck mode = DupCheck::Deny);
+    void doInsert(ColumnFamilyType type, TiKVKey && key, TiKVValue && value, DupCheck mode);
     void doCheckTable(const DecodedTiKVKey & key) const;
     void doRemove(ColumnFamilyType type, const TiKVKey & key);
 

